### PR TITLE
Sampling of the galaxy stellar masses

### PR DESCRIPTION
**Pull request by philipp128**
## Description
This module samples the galaxy stellar masses from a Schechter function according to the model of [1].
## Checklist
- [ ] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/master/CONTRIBUTING.md)
- [ ] Write unit tests
- [ ] Write documentation strings
- [ ] Assign someone from your working team to review this pull request
- [ ] Assign someone from the infrastructure team to review this pull request

## References
[1] Peng et al. (2010)

### CODE REVIEW

File: skypy/galaxy/tests/test_mass.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                              x_min=1e-10, x_max=q_max,
+                              resolution=1000)
+    d, p_value = scipy.stats.kstest(sample, 'truncexpon', args=(q_max,))
+    assert p_value >= 0.01
+
+
+def test_peng_masses():
+    # Test that error is returned if redshift input is an array but size !=
+    # None and size != redshift,size
+    with pytest.raises(ValueError):
```
Comment (philipp128) [inline skypy/galaxy/tests/test_mass.py:24]: This test is failing and I don't understand why. It's basically the same code as in the case of sampling the luminosities.

File: skypy/galaxy/mass.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+from skypy.utils.random import schechter
+
+
+__all__ = [
+    'peng_masses',
+]
+
+
+def peng_masses(redshift, alpha, m_star, size=None,
+                x_min=0.0002138, x_max=213.8, resolution=100):
```
Comment (rrjbca) [inline skypy/galaxy/mass.py:16]:
```suggestion
def peng_masses(alpha, m_star, size=None,
                x_min=0.0002138, x_max=213.8, resolution=100):
```

File: skypy/galaxy/mass.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    r"""Model of Peng et al (2010)
+
+    Stellar masses following the Schechter mass function following the
+    Peng et al. [1]_ model.
+
+    Parameters
+    ----------
+    redshift : (nz,) array-like
+        The redshift values at which to sample luminosities.
```
Comment (rrjbca) [inline skypy/galaxy/mass.py:25]:
```suggestion
```

File: skypy/galaxy/mass.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    alpha : float or int
+        The alpha parameter in the Schechter stellar mass function.
+    m_star : float or int
```
Comment (rrjbca) [inline skypy/galaxy/mass.py:28]:
```suggestion
    m_star : array-like
```

File: skypy/galaxy/mass.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        Characteristic stellar mass M_*.
+    size: int, optional
+         Output shape of stellar mass samples. If size is None and redshift
+         is a scalar, a single sample is returned. If size is None and
+         redshift is an array, an array of samples is returned with the same
+         shape as redshift.
```
Comment (rrjbca) [inline skypy/galaxy/mass.py:34]:
```suggestion
         Output shape of stellar mass samples. If size is None and m_star
         is a scalar, a single sample is returned. If size is None and
         m_star is an array, an array of samples is returned with the same
         shape as m_star.
```

File: skypy/galaxy/mass.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    >>> z = np.linspace(0,2, 100)
+    >>> masses = mass.peng_masses(z, -1.4, m_star = 10**10.67)
+
+
+
+    """
+
+    if size is None and np.shape(redshift):
+        size = np.shape(redshift)
```
Comment (rrjbca) [inline skypy/galaxy/mass.py:93]:
```suggestion
    if size is None and np.shape(m_star):
        size = np.shape(m_star)
```

File: skypy/galaxy/tests/test_mass.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        mass.peng_masses(np.array([1, 2]), -1.4, 10**10.67, size=3)
```
Comment (rrjbca) [inline skypy/galaxy/tests/test_mass.py:25]:
```suggestion
    # Test that error is returned if m_star input is an array but size !=
    # None and size != m_star.size
    with pytest.raises(ValueError):
        mass.peng_masses(-1.4, np.array([1e10, 2e10]), size=3)
```

**Comment by philipp128** [comment]
> There is no redshift dependence in this model so it doesn't make sense for this to be a function of redshift and the failing test is meaningless. However, we should generalise the function to accept m_star as an array. Then it would make sense for the function to raise an exception if size != m_star.size. I've given suggestions for how to make this change but you will need to update the rest of your examples and tests as well to make them consistent.

True. I am not sure whether it makes sense to make `m_star` array-like but not `alpha`?

**Comment by rrjbca** [comment]
This suggestion is equivalent to what we did for the luminosity function where L* was an array and alpha was constant. The only difference in that case was that instead of passing L* to the function directly we passed an array of redshifts and calculated L* at each redshift inside the function.

File: skypy/galaxy/mass.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
+r"""Models of galaxy luminosities.
```
Comment (ntessore) [inline skypy/galaxy/mass.py:1]:
```suggestion
r"""Models of galaxy mass (total and stellar).
```

If there will not be total galaxy mass in here, then `mass.py` is the wrong name for this file.

File: skypy/galaxy/mass.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+from skypy.utils.random import schechter
+
+
+__all__ = [
+    'peng_masses',
+]
+
+
+def peng_masses(redshift, alpha, m_star, size=None,
```
Comment (ntessore) [inline skypy/galaxy/mass.py:15]:
```suggestion
def stellar_mass_function(redshift, alpha, m_star, size=None,
```

This is the common Schechter stellar mass function as far as I can tell.

File: skypy/galaxy/mass.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+                x_min=0.0002138, x_max=213.8, resolution=100):
+    r"""Model of Peng et al (2010)
+
+    Stellar masses following the Schechter mass function following the
+    Peng et al. [1]_ model.
```
Comment (ntessore) [inline skypy/galaxy/mass.py:20]: Same, what's makes this the Peng et al. (2010) model?

File: skypy/galaxy/mass.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+    Returns
+    -------
+    stellar mass : array_like
+        Drawn stellar masses from the Schechter stellar mass function in units
+        of the solar mass
+
+    Notes
+    -----
+    The Schechter luminosity function is given as
```
Comment (ntessore) [inline skypy/galaxy/mass.py:48]:
```suggestion
    The stellar mass function is given by a Schechter function as
```

File: skypy/galaxy/mass.py (hunk 9 above)
Comment (philipp128) [reply skypy/galaxy/mass.py:15]: Ah ok. I just wrote Peng, cause that's where I saw it the first time.

File: skypy/galaxy/mass.py (hunk 8 above)
Comment (philipp128) [reply skypy/galaxy/mass.py:1]: Not yet. Maybe it comes in the future? I can change the name tho and if there will be total masses at some point, we can change it back.

Comment (rrjbca) [reply skypy/galaxy/mass.py:1]: The dark matter masses are handled in the halo module. So this should probably be stellar_mass.py or similar

Comment (ntessore) [reply skypy/galaxy/mass.py]: If the subpackage is `stellar_mass` then the function should probably be `schechter`. The other SMF I could think of is `broken_power_law`, so that `galaxy.stellar_mass.{schechter,broken_power_law}` makes more sense than `galaxy.stellar_mass.stellar_mass_function`.

Comment (philipp128) [reply skypy/galaxy/mass.py]: Ok this is true. I will change it then to this in the next commit.

Comment (philipp128) [reply skypy/galaxy/mass.py]: `schechter` already exists in `utils.random` and I am importing this. This gives an error with flake8 but also it's not really nice to have the same name for e more than one function in SkyPy, right? What do you think about `schechter_mass`?

Comment (Lucia-Fonseca) [reply skypy/galaxy/mass.py]: we might not need the Schechter function in `utils.random` after I push all the code in PR#186. We might need to discuss this offline at some point

Comment (rrjbca) [reply skypy/galaxy/mass.py]: I'm pretty sure we will want to keep `utils.random.schechter`. You could change how you import it to prevent the collision:
```
from skypy.utils import random
def schechter(...):
    x_sample = random.schechter(...)

Comment (philipp128) [reply skypy/galaxy/mass.py]: But is it nice to have more functions called 'schechter'? Also, it might be a bit misleading to call it just 'schechter', cause it's not a schechter function. Although it's in `stellar_mass.py` I don't really like it just calling schechter tbh.

**Comment by rrjbca** [comment]
We also need a test where `size = None` and m_star is an array to improve the test coverage. E.g. something that checks `m_star.shape == schechter_mass(..., m_star, ...).shape`

File: skypy/galaxy/mass.py (hunk 8 above)
Comment (ntessore) [reply skypy/galaxy/mass.py]: You can call it `schechter_smf`, and we can at some point rename the other function to `schechter_lf`. I think that both functions would ideally reside in the same module eventually, since they both deal with the luminous matter of a galaxy (eg. `skypy.galaxy.stars`) and there's little point in fracturing the module hierarchy more than that.

File: skypy/galaxy/stellar_mass.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        of the solar mass
+
+    Notes
+    -----
+     The stellar mass function is given by a Schechter function as
+
+    .. math::
+        \Phi(M, z) = \frac{\Phi_\star(z)}{M_\star}
+            \left(\frac{M}{M_\star}\right)^\alpha
+            \exp\left(-\frac{M}{M_\star}\right) \;.
```
Comment (ntessore) [inline skypy/galaxy/stellar_mass.py:47]: Not really relevant here since that is the number density, but the sampling does not draw galaxies from a Schechter function, it assigns masses to already drawn galaxies. This needs more explanation so that users understand what this function is doing and how it interacts with the actual sampling of galaxies, which currently happens in the redshift module.

File: skypy/galaxy/stellar_mass.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+from skypy.utils.random import schechter
+
+
+__all__ = [
+    'schechter_mass',
+]
+
+
+def schechter_mass(alpha, m_star, x_min, x_max, resolution=100, size=None):
```
Comment (ntessore) [inline skypy/galaxy/stellar_mass.py:15]: Confusing name since this is only stellar mass.

**Comment by ntessore** [comment]
Looks okay now, but the new module is not included in the documentation yet.

**Comment by philipp128** [comment]
> Looks okay now, but the new module is not included in the documentation yet.

How do you do this again? Which file is it?
